src/sqlfluff/rules/L003.py

```diff
@@ -1,15 +1,180 @@
 """Implementation of Rule L003."""
-from typing import List, Optional, Sequence, Tuple
+import dataclasses
+import itertools
+from typing import Dict, List, Optional, Sequence, Tuple
 
 from sqlfluff.core.parser import WhitespaceSegment
-from sqlfluff.core.parser.segments import BaseSegment, RawSegment
+from sqlfluff.core.parser.segments import BaseSegment
 from sqlfluff.core.rules.functional import rsp, sp, Segments
 from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
 from sqlfluff.core.rules.doc_decorators import (
     document_fix_compatible,
     document_configuration,
 )
 from sqlfluff.core.templaters import TemplatedFile
+from sqlfluff.core.templaters.base import RawFileSlice
+
+
+@dataclasses.dataclass
```

Comment:
Overall, big 👍 for adding these classes and moving some of the logic here.

Reply:
I like the classes but I think I'm a bit confused with the usage of the `_LineSummary` class - as you've mentioned later on it's both representation of data and also mutable.

From my first reading through - I feel like we have a generic `_LineSummary` class, but that we should have a subclass for a `_MemoryLine` so that in our type hints we can be really clear on how the summary is being used.

IMHO, I also think the naming of `memo_line` or `memory_line` is potentially misleading because it sounds similar to the `memory` of the rule and the (quite well thought through) `_Memory` class that already exists. I think calling it something like `reference_line` or `comparison_line` might distinguish it better.

Reply:
So... I went ahead and moved this one class to be a lot closer to immutable. We now generate fresh instances at each step. the concept of a Memory line is also most completely removed. Which is easier to reason about